CDAP-13280 Add Spark 2 streaming Kafka source

poornachandra opened the pull request:
JIRA - https://issues.cask.co/browse/CDAP-13280
Build - Ran `mvn clean package` locally

* Adds support for Spark 2 streaming Kafka source
  * Upgrade Kafka client for Spark 2 to 0.10.2.0, since Spark 2 needs 0.10 Kafka client for streaming
* Add test cases for Kafka streaming source, Kafka batch sink and Kafka alerts publisher
* Fix some dependencies in POM
* Update documentation
* Add screenshots and icons

Note: Majority of the change is due to renaming kafka-plugins-0.9 module to kafka-plugins-0.10

Commit:
CDAP-13280 Add Spark 2 streaming Kafka source

yaojiefeng commented on kafka-plugins-0.10/docs/KAFKASOURCE.md:
```diff
@@ -0,0 +1,85 @@
+
+## License and Trademarks
+
+Copyright © 2017 Cask Data, Inc.
```
This should be 2018

poornachandra replied:
Done

yaojiefeng commented on kafka-plugins-0.10/docs/Kafka-alert-publisher.md:
```diff
@@ -0,0 +1,63 @@
+
+## License and Trademarks
+
+Copyright © 2017 Cask Data, Inc.
```
This should be 2018

poornachandra replied:
Done

yaojiefeng commented on kafka-plugins-0.10/src/main/java/co/cask/hydrator/plugin/alertpublisher/KafkaAlertPublisher.java:
```diff
@@ -0,0 +1,174 @@
+        producer.send(record);
+      } catch (Exception e) {
+        // catch the exception and continue processing rest of the alerts
+        LOG.error("Exception while emitting alert {}", alert, e);
```
Should we limit the logs we sent, since we send the record one by one to kafka. If something breaks, this will flood the log

poornachandra replied:
Filed a JIRA to do this later since the log sampling classes are not directly usable now - https://issues.cask.co/browse/CDAP-13479

yaojiefeng commented on kafka-plugins-0.10/src/main/java/co/cask/hydrator/plugin/sink/Kafka.java:
```diff
@@ -33,10 +33,13 @@
 import co.cask.hydrator.common.KeyValueListParser;
 import co.cask.hydrator.common.ReferenceBatchSink;
 import co.cask.hydrator.common.ReferencePluginConfig;
+import co.cask.hydrator.plugin.common.KafkaHelpers;
```
It is not related to this pr, but I think we should name this class to KafkaBatchSink to be consistent with the KafkaBatchSource.

poornachandra replied:
Done

yaojiefeng commented on kafka-plugins-0.10/src/main/java/co/cask/hydrator/plugin/source/KafkaConfig.java:
```diff
@@ -0,0 +1,430 @@
+
+  public KafkaConfig(String referenceName, String brokers, String topic, String partitions,
+                     String initialPartitionOffsets, Long defaultInitialOffset, String schema, String format,
+                     String timeField, String keyField, String partitionField, String offsetField) {
```
I know this constructor is not used. But should we have principal and keytab location in the parameters since they are now fields in the config

poornachandra replied:
The constructor already has too many parameters, and the constructor is not used anywhere. So I just removed the constructors.

yaojiefeng commented on kafka-plugins-0.10/src/main/java/co/cask/hydrator/plugin/source/KafkaConfig.java:
```diff
@@ -0,0 +1,430 @@
+    return conf;
+  }
+
+  public void validate() {
```
Should we validate that the principal and keytab location can either be both null or non-null? I think we also  validate this in KafkaBatchSource, KafkaBatchSink and KafkaAlertPublisher

poornachandra replied:
Done

yaojiefeng commented on kafka-plugins-0.10/src/main/java/co/cask/hydrator/plugin/alertpublisher/KafkaAlertPublisher.java:
```diff
@@ -0,0 +1,174 @@
+        producer.send(record);
+      } catch (Exception e) {
+        // catch the exception and continue processing rest of the alerts
+        LOG.error("Exception while emitting alert {}", alert, e);
```
Should we limit the logs here since if something bad happens sending the alerts, the message may flood the log

poornachandra replied:
Duplicate comment

yaojiefeng commented on kafka-plugins-0.10/src/main/java/co/cask/hydrator/plugin/sink/Kafka.java:
```diff
@@ -33,10 +33,13 @@
 import org.apache.commons.lang3.StringUtils;
 import org.apache.hadoop.io.Text;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.common.serialization.StringSerializer;
```
Though it is not related to this pr, I think we should name this class as KafkaBatchSink to be consistent with the batch source, otherwise it will confuse people why this is called Kafka

poornachandra replied:
Done

yaojiefeng commented on kafka-plugins-0.10/src/main/java/co/cask/hydrator/plugin/source/KafkaStreamingSource.java:
```diff
@@ -0,0 +1,285 @@
+    kafkaParams.put("value.deserializer", ByteArrayDeserializer.class.getCanonicalName());
+    KafkaHelpers.setupKerberosLogin(kafkaParams, conf.getPrincipal(), conf.getKeytabLocation());
+    // Create a unique string for the group.id using the pipeline name and the topic
+    kafkaParams.put("group.id", Joiner.on("-").join(context.getPipelineName().length(), conf.getTopic().length(),
```
Can you add a comment about this property?

poornachandra replied:
Done

yaojiefeng commented on kafka-plugins-0.10/src/main/java/co/cask/hydrator/plugin/source/KafkaStreamingSource.java:
```diff
@@ -0,0 +1,285 @@
+      for (Map.Entry<TopicPartition, Long> entry : offsets.entrySet()) {
+        TopicPartition topicAndPartition = entry.getKey();
+        Long offset = entry.getValue();
+        if (offset == OffsetRequest.EarliestTime()) {
```
I don't understand this logic. The offsets can have -1 and -2 as their value, if that is the case, this will never match right?

poornachandra replied:
`OffsetRequest.EarliestTime()` always returns -2 and `OffsetRequest.LatestTime()` always returns -1, so they will match.

yaojiefeng replied:
👍

yaojiefeng commented on kafka-plugins-0.10/src/test/java/co/cask/hydrator/KafkaStreamingSourceTest.java:
```diff
@@ -0,0 +1,267 @@
+  @ClassRule
+  public static TemporaryFolder tmpFolder = new TemporaryFolder();
+
+
```
extra new line

poornachandra replied:
Done

yaojiefeng reviewed (commented):
Just some small comments. 
Also wondering if we should keep kafka 9 module, since this kafka 9 client will not be able to read from kafka 10

Commit:
Address review comments

poornachandra commented:
@yaojiefeng I have addressed the comments, please take another look.

Since not many people are using Kafka 0.9, I think it is expensive to have support for it. The Kafka 0.8 client can be still used for non-Kerberos Kafka 0.9 servers. Just that Kerberos support is not available. We can revisit this decision later if needed.

yaojiefeng commented on kafka-plugins-0.10/src/main/java/co/cask/hydrator/plugin/batchSource/KafkaBatchSource.java:
```diff
@@ -50,14 +50,9 @@
-import java.util.Set;
 import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.*;
```
We should use specific imports instead of using *

yaojiefeng commented on kafka-plugins-0.10/src/main/java/co/cask/hydrator/plugin/source/KafkaConfig.java:
```diff
@@ -24,21 +24,16 @@
-import java.util.Map;
-import java.util.Set;
-import javax.annotation.Nullable;
+import java.util.*;
```
We should use specific imports instead of using *

yaojiefeng commented on kafka-plugins-0.10/src/main/java/co/cask/hydrator/plugin/source/KafkaStreamingSource.java:
```diff
@@ -52,14 +52,7 @@
-import java.util.Map;
-import java.util.Properties;
-import java.util.Set;
+import java.util.*;
```
We should use specific imports instead of using *

yaojiefeng commented on kafka-plugins-0.10/src/test/java/co/cask/hydrator/KafkaSinkAndAlertsPublisherTest.java:
```diff
@@ -55,21 +55,10 @@
-import org.junit.BeforeClass;
-import org.junit.ClassRule;
-import org.junit.Test;
+import org.junit.*;
```
We should use specific imports instead of using *

yaojiefeng reviewed (approved):
Just a few comments about import, rest LGTM

Commit:
Add SASL config to Kafka output format

Commit:
Fix imports

Commit:
Replace Kafka constant to prevent class not found exception

yaojiefeng reviewed (approved):
lgtm
